src/mame/sinclair/spec128.cpp

```diff
@@ -288,31 +288,106 @@ void spectrum_128_state::spectrum_128_fetch(address_map &map)
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Cursor Up") PORT_CODE(KEYCODE_UP)
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Cursor Down") PORT_CODE(KEYCODE_DOWN)
+	PORT_BIT(0xe0, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("PLUS2") // Spectrum+ Keys (Same as CAPS + SPACE and CAPS + SYMBOL)
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("BREAK") PORT_CODE(KEYCODE_PAUSE)
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("EXT MODE") PORT_CODE(KEYCODE_LCONTROL)
+	PORT_BIT(0xfc, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("PLUS3") // Spectrum+ Keys (Same as SYMBOL SHIFT + O/P)
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("\"") PORT_CODE(KEYCODE_QUOTE)
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME(";") PORT_CODE(KEYCODE_COLON)
+	PORT_BIT(0xfc, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("PLUS4") // Spectrum+ Keys (Same as SYMBOL SHIFT + N/M)
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME(".") PORT_CODE(KEYCODE_STOP)
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME(",") PORT_CODE(KEYCODE_COMMA)
+	PORT_BIT(0xf3, IP_ACTIVE_LOW, IPT_UNUSED)
+INPUT_PORTS_END
+
+INPUT_PORTS_START( spec_plus )
+	PORT_INCLUDE( spec128 )
+	PORT_INCLUDE( spec_plus_joys )
+
+	PORT_START("MOD_DMA")
+	PORT_CONFNAME( 0x03, 0x00, "DMA")
+	PORT_CONFSETTING(    0x00, DEF_STR( No ))
+	//PORT_CONFSETTING(    0x01, "UA858D")
+	PORT_CONFSETTING(    0x03, "Zilog")
+	PORT_CONFNAME( 0x04, 0x00, "DMA Port") PORT_CONDITION("MOD_DMA", 0x01, EQUALS, 0x01)
+	PORT_CONFSETTING(    0x00, "11: MB02+")
+	PORT_CONFSETTING(    0x04, "107: DATAGEAR")
+INPUT_PORTS_END
+
 void spectrum_128_state::machine_start()
 {
 	spectrum_state::machine_start();
 
 	save_item(NAME(m_port_7ffd_data));
 
-	/* rom 0 is 128K rom, rom 1 is 48 BASIC */
+	// rom 0 is 128K rom, rom 1 is 48 BASIC
 	memory_region *rom = memregion("maincpu");
 	m_bank_rom[0]->configure_entries(0, 2, rom->base() + 0x10000, 0x4000);
 
 	auto ram_entries = m_ram->size() / 0x4000;
 	for (auto i = 1; i < 4; i++)
 		m_bank_ram[i]->configure_entries(0, ram_entries, m_ram->pointer(), 0x4000);
 
-	m_bank_ram[1]->set_entry(ram_entries > 5 ? 5 : (ram_entries - 1)); /* Bank 5 is always in 0x4000 - 0x7fff */
-	m_bank_ram[2]->set_entry(2); /* Bank 2 is always in 0x8000 - 0xbfff */
+	m_bank_ram[1]->set_entry(ram_entries > 5 ? 5 : (ram_entries - 1)); // Bank 5 is always in 0x4000 - 0x7fff
+	m_bank_ram[2]->set_entry(2); // Bank 2 is always in 0x8000 - 0xbfff
 }
 
 void spectrum_128_state::machine_reset()
 {
 	spectrum_state::machine_reset();
 
-	/* set initial ram config */
+	// set initial ram config
 	m_port_7ffd_data = 0;
 	spectrum_128_update_memory();
+
+	const u8 mod_dma = m_mod_dma.read_safe(0);
+	if (mod_dma & 1)
+	{
+		// mod_dma & 2 ? ZILOG : UA858D
+		const u8 port = mod_dma & 4 ? 0x6b : 0x0b;
+		m_maincpu->space(AS_IO).install_readwrite_handler(port, port, 0, 0xff00, 0, read8smo_delegate(*m_dma, FUNC(z80dma_device::read)), write8smo_delegate(*m_dma, FUNC(z80dma_device::write)));
+	}
```

Comment:
What’s going to happen if the user changes the configuration setting and then soft resets the emulated system?  Will this end up installed in two places?  What if they disable it and reset?  Will it still be installed from the previous start?

If there’s no clean way to uninstall this if the configuration changes before a soft reset, you need to only install it on the first reset.  A number of ISA cards, etc. do this check.